### ISSUE TITLE
Interest in Unit Dual Quaternions?

### ISSUE THREAD

**Issue by artivis** [artivis/manif#164]
After seeing [this post on ROS Dicourse](https://discourse.ros.org/t/release-of-dual-quaternions/15911) and the comments asking for a c++ implementation, I wondered if 'Unit Dual Quaternions' (`DH_u` I believe) would be an interesting/useful addition to `manif` given that it is somewhat equivalent to `SE3`.

**Comment by joansola** [comment]
I agree. Just for the fun of it, and just because we could be the firsts to have it!

I however dont know how it works

But lets check it out

**Comment by joansola** [comment]
See appendix in https://cs.gmu.edu/~jmlien/teaching/cs451/uploads/Main/dual-quaternion.pdf for arithmetic.

Exp and Log are not in the paper. Neither the tangent space.

**Comment by artivis** [comment]
Alright then I'll get started with the class layout and we can fill up the implementation as we uncover it.

**Comment by joansola** [comment]
Would this constitute a replacement for SE3? It seems that's the case! Or
sould we rather keep both?

**Comment by artivis** [comment]
No we should definitely keep both. `DHu` would be an alternative to `SE3` for whomever wants to use it for whatever reason...

**Comment by prashanthr05** [comment]
Related library from the ROS discourse https://github.com/Achllle/dual_quaternions 

This could be really interesting. I am looking forward to this feature.

**Comment by joansola** [comment]
I believe this is a relevant relative https://dqrobotics.github.io/

**Comment by joansola** [comment]
I got in contact with Bruno Adorno @bvadorno @dqrobotics who is a specialist on dual quaternions.

**Comment by artivis** [comment]
Good to hear! Is it something you want to pick up? I don't know when I'll have time to resume this.

**Comment by bvadorno** [comment]
Hi @joansola and @artivis!

It'd be my pleasure to help you guys in this endeavor. I've followed the discussion trail and there are some things that may be helpful to you.

DQ Robotics is implemented in C++, Matlab, and Python (actually we use bindings to access the C++ implementation, so on the high level the user sees Python code, but under the hood, everything runs in C++, which makes the Python code suitable for realtime applications).

Several operations that I see you guys need are already there: exponential and logarithmic maps, different Jacobians. I still have to see with @joansola if there is a Jacobian that matches the Jacobian you guys use. Probably yes, but since the library implements the general operations of *dual quaternion algebra* and *dual quaternion calculus*, only a subset of it is related to Lie Algebra, namely the subset of *unit* dual quaternions. Therefore, we still need to see which Jacobian is the one you guys usually work with.

We have recently published a paper showcasing the main points of the library that may be useful: 

Adorno, Bruno Vilhena, and Murilo Marques Marinho. 2020. “DQ Robotics: A Library for Robot Modeling and Control.” IEEE Robotics & Automation Magazine, 0–0. https://doi.org/10.1109/MRA.2020.2997920. There is a version available on [ArXiv](http://arxiv.org/abs/1910.11612).

The overall notation is described in something that I've been writing for a while and hopefully will become a book soon: Adorno, Bruno Vilhena. 2017. “Robot Kinematic Modeling and Control Based on Dual Quaternion Algebra -- Part I: Fundamentals.” You can download it [here](https://hal.archives-ouvertes.fr/hal-01478225v1). The idea is to have three parts, and the first one is related to the fundamentals of dual quaternion algebra. It includes several elements that you guys often use in Lie Theory (the concept of groups, Lie Algebra, tangent space, etc.), but I tried to keep the jargon to a minimum so the material could be self-contained.

There are several things implemented in the library that are scattered around within the papers I  co-authored, but they are usually cited in the library documentation. Whenever you get the chance to take a look at it, if there's anything I can do to help, please let me know.

Cheers,
Bruno

**Comment by joansola** [comment]
Hi Bruno welcome to the fun!!

**Comment by artivis** [comment]
Hi @bvadorno,
Thanks a lot for the references! We will be looking into them :+1:
